### PR TITLE
fix(gridSystem): fix `itemsPerRow` precision

### PR DESCRIPTION
**Pull request by psychedelicious**
- Fix `itemsPerRow()` calculation to handle edge cases related to element dimension precision
- Add tests with experimentally-determined edge cases that cause precision problems with the
previous calculation
- fixes [BUG] `VirtuosoGrid` + CSS Grid `List` component: items not rendered #889

`react-virtuoso` calculates element dimensions using `getBoundingClientRect()`. This value may be
imprecise when the browser view is zoomed in/out, though technically it is [more precise than eg
`clientWidth`](https://developer.mozilla.org/en-US/docs/Web/API/Element/clientWidth).

To see this for yourself, use `getBoundingClientWidth()` and `clientWidth` to measure the width of
a relatively sized element at various zoom levels. `clientWidth` reports the same value each time,
while `getBoundingClientRect()` may report slightly different values.

This very small difference may be magnified during the calculation done in `itemsPerRow()` and, in
certain cases, causes the resulting value to be 1 less than what it should be. The number of items
to render in the window are thus incorrect.

My crude understanding of the cause: when the browser calculates viewport and element
dimensions at non-100% zoom level, some native rounding is involved. `getBoundingClientRect()`,
which reports precise fractional values, is affected by this rounding, while `clientWidth` rounds or
floors values. This may not be entirely correct - discussion on this is rather sparse and anyways,
it's beyond me.

### CODE REVIEW

**Comment by psychedelicious** [comment]
Side-note: `urx` is really cool!

File: src/gridSystem.ts [hunk 1]
```diff
@@ -349,6 +349,6 @@ function itemTop(viewport: ElementDimensions, gap: Gap, item: ElementDimensions,
   return top > 0 ? top + gap.row : top
 }
 
-function itemsPerRow(viewportWidth: number, itemWidth: number, gap: number) {
-  return max(1, floor((viewportWidth + gap) / (itemWidth + gap)))
+export function itemsPerRow(viewportWidth: number, itemWidth: number, gap: number) {
+  return max(1, floor((floor(viewportWidth) + gap) / (floor(itemWidth) + gap)))
```
Comment (petyosi) [inline src/gridSystem.ts:353]: Question - wouldn't it be more conservative to use `ceil(itemWidth)` instead? I understand that we're basically trying to guess if the browser will fit those in there...

**Comment by petyosi** [comment]
I hear you - this is most likely the problem... Can this, however, cause the opposite kind of mistake? As in, calculate that 4 items can fit the row, while in reality, only 3 fit?

File: src/gridSystem.ts (hunk 1 above)
Comment (psychedelicious) [reply src/gridSystem.ts]: Actually, `ceil` is less conservative, because `itemWidth` is the divisor - `ceil`ing `itemWidth` can only result in a lesser `itemsPerRow` (and indeed, using `ceil` does not fix the problem).

The `floor(viewportWidth)` is, I believe, not needed. I cannot find a case where the calculation is different without it, but removing that `floor` does make this a more conservative calculation (erring on the side of rendering too many items in the rare edges cases). I've removed that `floor` call.

**Comment by psychedelicious** [comment]
I think it is possible in some very rare edge case, but during testing of a wide range of relatively sized grid configurations, I did not see it calculating more items than it should.

I assume it is better to calculate too large of a number than too small. Too large and the change in behaviour is a performance degradation. To small and the change in behaviour is a problem with data display.
